storage-node/src/cache/data_store_cache/memdisk/cache_manager.rs

```diff
@@ -0,0 +1,361 @@
+        let data_size = self
+            .disk_store
+            .write_data(disk_store_key.clone(), bytes_to_disk, Some(data_stream))
+            .await?;
+        let mut disk_replacer = self.disk_replacer.write().await;
+        if disk_replacer
+            .put(remote_location, (disk_store_key.clone(), data_size))
+            .is_none()
+        {
+            self.disk_store.clean_data(&disk_store_key).await?;
+            // TODO: do we need to notify the caller this failure?
+        }
+        Ok(data_size)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        cache::policy::lru::LruReplacer,
+        storage_reader::{s3_diskmock::MockS3Reader, AsyncStorageReader},
+    };
+
+    use super::*;
+    #[tokio::test]
+    async fn test_put_data_evicted_from_mem() {
+        // 1. put small data1
+        // 2. put small data2, making data1 evicted
+        //    (TODO(lanlou): when adding pin & unpin, this test should be modified)
+        // 3. get data1, should be from disk
+        // 4. get data2, should be from memory
+        // TODO(lanlou): What if we continue accessing data1 and never accessing data2 in
+        // the future? There is no chance for data1 to be put back to memory again currently.
+        let tmp = tempfile::tempdir().unwrap();
+        let disk_base_path = tmp.path().to_owned();
+        let mut cache = MemDiskStoreCache::new(
+            LruReplacer::new(1024 * 512),
+            disk_base_path.to_str().unwrap().to_string(),
+            Some(LruReplacer::new(950)),
+            None,
+        );
+        let bucket = "tests-text".to_string();
+        let keys = vec!["what-can-i-hold-you-with".to_string()];
+        let data1_key = bucket.clone() + &keys[0];
+        let reader = MockS3Reader::new(bucket.clone(), keys.clone()).await;
+        let mut data_size = cache
+            .put_data_to_cache(data1_key.clone(), reader.into_stream().await.unwrap())
+            .await
+            .unwrap();
+        assert_eq!(data_size, 930);
+
+        let data2_key = bucket.clone() + &keys[0] + "1";
+        let reader = MockS3Reader::new(bucket.clone(), keys).await;
+        data_size = cache
+            .put_data_to_cache(data2_key.clone(), reader.into_stream().await.unwrap())
+            .await
+            .unwrap();
+        assert_eq!(data_size, 930);
+
+        data_size = 0;
+        let mut rx = cache.get_data_from_cache(data1_key).await.unwrap().unwrap();
+        while let Some(data) = rx.recv().await {
+            let data = data.unwrap();
+            data_size += data.len();
+        }
+        assert_eq!(data_size, 930);
+
+        data_size = 0;
+        let mut rx = cache.get_data_from_cache(data2_key).await.unwrap().unwrap();
+        while let Some(data) = rx.recv().await {
+            let data = data.unwrap();
+            data_size += data.len();
+        }
+        assert_eq!(data_size, 930);
+    }
+
+    #[tokio::test]
+    async fn test_mixed_disk_only() {
```

Comment:
Why do we call it "mixed" disk only? 👀

Reply:
Oh it means put and get. I will change the name.